test-harness/src/checkpoint.rs

```diff
@@ -0,0 +1,104 @@
+use std::fmt::Debug;
+use std::fmt::Display;
+use std::ops::Range;
+use std::ops::RangeFrom;
+use std::ops::RangeFull;
+use std::ops::RangeInclusive;
+use std::ops::RangeTo;
+use std::ops::RangeToInclusive;
+use std::slice::SliceIndex;
+
+use crate::Event;
+
+/// A checkpoint in a `ghcid-ng` run.
+///
+/// [`crate::GhcidNg`] provides methods for asserting that events are logged, or waiting for
+/// events to be logged in the future.
+///
+/// To avoid searching thousands of log events for each assertion, and to provide greater
+/// granularity for assertions, you can additionally assert that events are logged between
+/// particular checkpoints.
+///
+/// Checkpoints can be constructed with [`crate::GhcidNg::first_checkpoint`],
+/// [`crate::GhcidNg::current_checkpoint`], and [`crate::GhcidNg::checkpoint`].
+#[derive(Debug, Clone, Copy)]
+pub struct Checkpoint(pub(crate) usize);
+
+impl Checkpoint {
+    /// Convert this checkpoint into an index.
+    pub fn into_index(self) -> usize {
```

Comment:
It confused me quite a bit that this `into_index` method on `Checkpoint` had the same name as the `into_index` method on `CheckpointIndex`.  Maybe give this method a different name or just access the `.usize` field directly instead of using this method since that field is already crate-visible